src/OrbitGl/CallstackThreadBar.cpp

```diff
@@ -226,60 +226,116 @@ bool CallstackThreadBar::IsEmpty() const {
+  const std::string& function_name = module_and_function_name.function_name;
+  const std::string shortened_function_name =
+      (max_length > 0)
+          ? orbit_gl::ShortenStringWithEllipsis(function_name, static_cast<size_t>(max_length))
           : function_name;
-  // Simple HTML escaping
-  return absl::StrReplaceAll(fn_name, {{"&", "&amp;"}, {"<", "&lt;"}, {">", "&gt;"}});
+  // Simple HTML escaping.
+  const std::string escaped_function_name =
+      absl::StrReplaceAll(shortened_function_name, {{"&", "&amp;"}, {"<", "&lt;"}, {">", "&gt;"}});
+  return module_and_function_name.function_is_unknown
+             ? absl::StrFormat("<i>%s</i>", escaped_function_name)
+             : escaped_function_name;
 }
 
-constexpr const char* kErrorColorString = "#ffb000";
+constexpr const char* kUnwindErrorColorString = "#ffb000";
 
-std::string CallstackThreadBar::FormatCallstackForTooltip(const CallstackInfo& callstack,
-                                                          int max_line_length, int max_lines,
-                                                          int bottom_n_lines) const {
-  std::string result;
-  int size = static_cast<int>(callstack.frames().size());
-  if (max_lines <= 0) {
-    max_lines = size;
-  }
-  const int bottom_n = std::min(std::min(max_lines - 1, bottom_n_lines), size);
-  const int top_n = std::min(max_lines, size) - bottom_n;
+std::string CallstackThreadBar::FormatCallstackForTooltip(const CallstackInfo& callstack) const {
+  constexpr int kMaxLineLength = 100;
+  constexpr int kMaxLines = 20;
+  constexpr int kBottomLineCount = 5;
+
+  int callstack_size = static_cast<int>(callstack.frames().size());
+  const int bottom_n = std::min(std::min(kMaxLines - 1, kBottomLineCount), callstack_size);
+  const int top_n = std::min(kMaxLines, callstack_size) - bottom_n;
 
+  constexpr int kShortenedForReadabilityFakeIndex = -1;
+  std::vector<int> frame_indices_to_display;
   for (int i = 0; i < top_n; ++i) {
-    // The first frame is always correct.
-    if (callstack.IsUnwindingError() && i > 0) {
-      result.append(absl::StrFormat("<br/><span style=\" color:%s;\">%s</span>", kErrorColorString,
-                                    SafeGetFormattedFunctionName(callstack, i, max_line_length)));
-    } else {
-      result.append("<br/>" + SafeGetFormattedFunctionName(callstack, i, max_line_length));
-    }
+    frame_indices_to_display.push_back(i);
   }
-  if (max_lines < size) {
-    result += "<br/><i>... shortened for readability ...</i>";
+  if (kMaxLines < callstack_size) {
+    frame_indices_to_display.push_back(-1);
   }
-  for (int i = size - bottom_n; i < size; ++i) {
+  for (int i = callstack_size - bottom_n; i < callstack_size; ++i) {
+    frame_indices_to_display.push_back(i);
+  }
+
+  std::string result;
+  for (int frame_index : frame_indices_to_display) {
+    if (frame_index == kShortenedForReadabilityFakeIndex) {
+      result += "<i>... shortened for readability ...</i><br/>";
```

Comment:
Nit: Since you're introducing constants already, the text could be moved into a constant as well.